### ISSUE TITLE
Jax support

### ISSUE THREAD

**Issue by bamos** [cvxgrp/cvxpylayers#7]
This would broadly involve:

1. Adding a `jax` directory that follows the same structure as the PyTorch/Tensorflow ones to our main module and appropriately add that layer: https://github.com/cvxgrp/cvxpylayers/tree/master/cvxpylayers
2. Filling in the details of the module and tests
3. Updating the README/image to include Jax
4. Adding some Jax examples: https://github.com/cvxgrp/cvxpylayers/tree/master/examples

\cc @gsp-27 who has expressed interest in helping with this -- please let us know if you make any progress here!

**Comment by joaogui1** [comment]
I would like to try, is that ok?

**Comment by bamos** [comment]
> I would like to try, is that ok?

Yes! Maybe you or @gsp-27 can take a lock on the initial version, send in a PR, and we can all coordinate afterwards? What do you all think?

**Comment by joaogui1** [comment]
> > I would like to try, is that ok?
> 
> Yes! Maybe you or @gsp-27 can take a lock on the initial version, send in a PR, and we can all coordinate afterwards? What do you all think?

Sounds great

**Comment by gsp-27** [comment]
Hi @joaogui1, sure give it a shot. I can also help.

Thanks,
Gaurav Pathak

**Comment by rodrigob** [comment]
any update on this ?

**Comment by joaogui1** [comment]
I didn't manage to do it, sorry

**Comment by rodrigob** [comment]
lack of time or too high technical difficulty ?

**Comment by joaogui1** [comment]
High technical difficulty, don't understand enough about CVXPY to understand the errors

**Comment by sbarratt** [comment]
I'm taking a crack at this

**Comment by akshayka** [comment]
@sbarratt, you should be able to use 

https://jax.readthedocs.io/en/latest/jax.html#jax.custom_gradient

to define the gradient.

**Comment by sbarratt** [comment]
That doesn't work. It still tries to trace the function. We need
https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html
to define a custom non-traceable python function.

**Comment by bodono** [comment]
Are there any updates on this?

**Comment by bamos** [comment]
I've been recently thinking about adding a jax/flax version but haven't started.

**Comment by bamos** [comment]
@bodono do you (or anyone else) have any recommendations/preferences on what a JAX implementation would look like? Maybe it would make the most sense to have a pure JAX version that is easily wrappable with at least flax/haiku? This way we could (hopefully) easily connect in any future JAX libraries/frameworks

**Comment by bodono** [comment]
I don't feel strongly about it and I'm very far from a jax expert, but I agree that having a pure jax version would probably be best.

**Comment by eserie** [comment]
Hi all, did you considered to use [eagerpy](https://github.com/jonasrauber/eagerpy)  to factorize the three tensor implementations (tensorflow, pytorch and jax)?

**Comment by bamos** [comment]
> Hi all, did you considered to use [eagerpy](https://github.com/jonasrauber/eagerpy) to factorize the three tensor implementations (tensorflow, pytorch and jax)?

That looks useful! Here we'd need to define new operations in each of the frameworks with custom derivatives and I don't immediately see how to do that in the eagerpy docs

**Comment by bamos** [comment]
Also this issue is resolved with #100

**Comment by eserie** [comment]
> That looks useful! Here we'd need to define new operations in each of the frameworks with custom derivatives and I don't immediately see how to do that in the eagerpy docs

Yes sure, its true that eagerpy is more design to go in the other way around: write some code that uses a tensor framework to compute derivatives of it. So not sure it could help here. However, maybe the fact to have lightweighted unified tensor API may help to factorize a bit.

I'll have a look at the Jax integration in #100!
